### PR TITLE
Added API support for sending messages to groups

### PR DESCRIPTION
**Pull request by rowanseymour**
Needs this for ChatPro. Also cleaned up docs for the messages endpoint.

### CODE REVIEW

File: temba/api/serializers.py [hunk 1]
```diff
@@ -1005,14 +1006,6 @@ def __init__(self, *args, **kwargs):
 
         super(MsgCreateSerializer, self).__init__(*args, **kwargs)
 
-    def validate(self, attrs):
-        urns = attrs.get('urn', [])
-        phone = attrs.get('phone', None)
-        contact = attrs.get('contact', [])
-        if (not urns and not phone and not contact) or (urns and phone):
-            raise ValidationError("Must provide either urns or phone or contact and not both")
```
Comment (nicpottier) [inline temba/api/serializers.py]: Is this case still covered? (no group, phone, contact or urn?)

File: temba/api/serializers.py [hunk 2]
```diff
@@ -1083,34 +1079,47 @@ def validate_phone(self, attrs, source):
...
+        groups = attrs.get('group', [])
+        for group in groups:
+            recipients.append(group)
 
         # create the broadcast
-        broadcast = Broadcast.create(org, user, attrs['text'], recipients=contacts)
+        broadcast = Broadcast.create(org, user, attrs['text'], recipients=recipients)
 
         # send it
         broadcast.send()
```
Comment (nicpottier) [inline temba/api/serializers.py]: Now that this is going against groups, seems this send could take a long time. (say we are sending to all UReporters)

Perhaps this should be firing a celery task instead? (```send_broadcast_task```)

Comment (rowanseymour) [inline temba/api/serializers.py]: Yeah that makes sense. Will update.

Actually I forgot to ask about this: documentation for this endpoint says that calls are limited to 100 URNs / 100 phone numbers - but there's no logic here to enforce that. I'm thinking that limit doesn't make so much sense if you can send to a group with 100000 contacts in it, and so we should update the docs...

Comment (nicpottier) [inline temba/api/serializers.py]: Are you still thinking of adding a broadcasts endpoint?

Wondering whether maybe we should add the group stuff to that one instead. The reason being is that the interface for the Messages endpoint is meant to return the id's of the messages created. In the case of a large group broadcast, those would be done in Celery, so we wouldn't actually have them.

If we had a ```broadcast``` endpoint, then we could allow writing there (with the addition of groups as you have here) and instead return a broadcast id that could be polled instead..

thoughts?

Comment (rowanseymour) [inline temba/api/serializers.py]: I wasn't but that would appear to be the only way to delay the .send() call. I'll add it without API documentation for now, and if all is good, we can deprecate POSTing to messages.json and update the docs.

File: temba/api/serializers.py [hunk 3]
```diff
@@ -978,15 +977,117 @@ def initialize(self, parent, field_name):
...
+
+    def get_contacts(self, obj):
+        return [contact.uuid for contact in obj.contacts.all()]
+
+    def get_groups(self, obj):
+        return [group.uuid for group in obj.groups.all()]
 
     def get_messages(self, obj):
-        return [msg.id for msg in obj.get_messages()]
+        if obj.status == INITIALIZING:
```
Comment (rowanseymour) [inline temba/api/serializers.py]: So we're not trying to read from messages whilst celery task is saving messages

File: temba/api/serializers.py [hunk 4]
```diff
@@ -315,7 +314,7 @@ def restore_object(self, attrs, instance=None):
         if uuid:
             contact.update_urns(urns)
         else:
-            contact = Contact.get_or_create(self.user, org, urns=urns, uuid=uuid)
+            contact = Contact.get_or_create(org, self.user, urns=urns, uuid=uuid)
```
Comment (rowanseymour) [inline temba/api/serializers.py]: Figured might as well make this follow convention (org, user, ...) whilst I was checking all calls to it

File: temba/contacts/models.py [hunk 5]
```diff
@@ -331,29 +331,22 @@ def from_urn(cls, org, scheme, path, country=None):
         return existing[0].contact if existing else None
 
     @classmethod
-    def get_or_create(cls, user, org, name=None, urns=None, channel=None, uuid=None):
+    def get_or_create(cls, org, user, name=None, urns=None, incoming_channel=None, uuid=None):
```
Comment (rowanseymour) [inline temba/contacts/models.py]: Figured might as well make this follow convention (org, user, ...) whilst I was checking all calls to it

File: temba/api/serializers.py [hunk 6]
```diff
@@ -978,15 +977,128 @@ def initialize(self, parent, field_name):
...
-        fields = ('messages', 'sms')
+        fields = ('id', 'urns', 'contacts', 'groups', 'text', 'messages', 'created_on', 'status')
+
+
+class BroadcastCreateSerializer(serializers.Serializer):
+    urns = StringArrayField(required=False)
+    contacts = StringArrayField(required=False)
+    groups = StringArrayField(required=False)
+    text = serializers.CharField(required=True, max_length=480)
+    channel = ChannelField(queryset=Channel.objects.filter(pk=-1), required=False)
```
Comment (rowanseymour) [inline temba/api/serializers.py]: Ideally channel would be fetched by UUID, but we have to tidy up our usage of `Channel.uuid` before we can do that

Comment (nicpottier) [inline temba/api/serializers.py]: Ya, not sure channels are 'forever' anyways, so having an integer id makes sense here.

Comment (ericnewcomer) [inline temba/api/serializers.py]: I would describe 'forever' as things that we will never remove on the user's behalf. So I would think channel would fit that description.

Comment (nicpottier) [inline temba/api/serializers.py]: True.

File: temba/api/views.py [hunk 7]
```diff
@@ -424,7 +591,7 @@ class MessagesEndpoint(generics.ListAPIView):
 
     Returns the message activity for your organization, listing the most recent messages first.
 
-      * **relayer** - the id of the Android channel that sent or received this message (int) (filterable: ```channel```)
+      * **channel** - the id of the Android channel that sent or received this message (int) (filterable: ```channel```)
       * **urn** - the URN of the sender or receiver, depending on direction (string) (filterable: ```urn```)
       * **contact** - the UUID of the contact (string) (filterable: ```contact```)
       * **group_uuids** - the UUIDs of any groups the contact belongs to (string) (filterable: ```group_uuids```)
```
Comment (ericnewcomer) [inline temba/api/views.py]: This ```group_uuids``` looks to be missing from ```get_read_explorer```

**Comment by nicpottier** [comment]
Looks good to me.

**Pull request merged**